### ISSUE TITLE
[3.2] Cooperative Mode - Discussion? 

### ISSUE THREAD

**Issue by Tormentor667** [Realm667/WolfenDoom#623]
I got a few messages already regarding the cooperative mode for Blade of Agony. I wonder, would it be possible? While mapping, I always placed 4 player startes and also made sure to create places in a way that makes it coop-compatible. I am not sure about the code-side though.

**Comment by Ozymandias81** [comment]
Well many maps wouldn't be possible in coop actually, such as Intermap, C1M6, C1M0, C2M4. C2M1, C3M1, C3INTRO, the blimpship, Keen maps, Castle Wolfenstein probably since there are many things involved that are single player (disguise uniforms, sneakable actors behavior, radio messages, maybe issues with PPShaders). A lot has to be done, and well BoA would be possible to be played coop only under GZDoom, which I am not sure if it is good for that.

**Comment by Tormentor667** [comment]
How can you be that sure that the Intermap or C1M6 wouldn’t work?

**Comment by AFADoomer** [comment]
I can't think of anything major that won't work...  some aesthetic things may need to be reworked (e.g., static skyboxes), and some map scripts weren't written with multi-player in mind. Disguises and stealth should still work, but one player being discovered will affect all players in the area.

**Comment by Tormentor667** [comment]
What we could consider is trying this as soon as we have take care of the things that need to be done. I can do a full playtest of all 3 episodes at home with my girlfriend and check how the maps are working - try to break things for example.

There is also a question of balancing the mod when it comes to treasure. Do all players share the same money for example? Do we multiply that by a certain factor?

Regarding NPCs, which player are they following?

**Comment by AFADoomer** [comment]
Whichever one they see first.  They iterate from player 1 to player 8 and see who they can see.  All should be able to activate/deactivate them, though.

**Comment by Tormentor667** [comment]
Regarding "connecting players", as of now there is no GZDoom interface but the console commands. Would it be possible to write some ZScript menue where players can "host" and "join" (via IP)?

**Comment by Ozymandias81** [comment]
Because I play under Zandronum quite every night since 2014, coop wads mainly, and the most complex one has been Stronghold in my eyes. Here with BoA there are way to many things happening on player side, also returning to HQs will mean that players which aren't 1st will be placed on different zones at respawn, and what if a player will have all Eisenmann files and others not? What will happen also at training session? Money won't be shared, objectives such as treasure chests will be good just for 1 player only. Then we have railshooter sequences, such as c1m6 and c2m4, and imagine what could happen with tanks too. I just think that to implement these things we will open an infinite amount of issues again, idk how to code them okay and maybe I am wrong, but not on the issue side and the fact that BoA is only GZDoom compatible. This imho could happen only once we are 100% sure that everything works fine and that we are using a stable release of GZDoom.

**Comment by Tormentor667** [comment]
I think most of the issues could be addressed if we implement a shared inventory - meaning that items, puzzle objects and other things are shared between all players.

If one player collects an Eisenmann file, all players have it. If a player gives a file to Gutenberg, it gets removed from everyone. That way, a lot of issues could be handled.

The objective scripts are global anyway, so if one player finishes an objective, it is finished for everyone. We‘d only make sure that the messages are printed for everyone and the sounds are played for everyone which is easily doable.

For the railshooting sequences we could either add all players to the sequence or make all players who do not take the ride give the opportunity to teleport to the ride‘s destination after a while.

I remember that I even added something like this to the ending battle of C2M2, having coop in mind.

Regarding the script parts, maybe @Talon1024 and @AFADoomer can say more about that. 

This would be definitely something for a 3.2 release.

**Comment by DoomJedi** [comment]
We have 1dir Bosses and some enemies.

...what about fps?
Can it be splitscreen too?

**Comment by Tormentor667** [comment]
Won’t matter regarding fps. 1dir bosses is a thing but maybe we can come up with a clever trick instead of you having to draw full rotations for all bosses 😅

**Comment by DoomJedi** [comment]
I hope you will :) Not that I have ideas (beside scaling 64x64 Bosses in coop mode, but it'll work (without extra work) only for classic ones, and what about brightmaps etc....)

As a maker of "SplitWolf" (coop splitscreen Wolf3D mod, for Windows and for Retropie) - I have alot of experience on coop modes, related challenges and even coop game modes ideas, some used as BoA mutators.
That's why I immediately mentioned 1-dir Bosses and enemies in particular, I was lucky that for Wolf3D in 64x64 - Boss work was already done by Untrustable. This is not the case with MACres Bosses and is simply impossible work at this point and in general.

**Comment by DoomJedi** [comment]
Another question - how enemy AI work in such cases? Will they attack closest player at sight, or some less predictable pattern?
Also the mod can be too easy for coop play.
In SplitWolf we solved that by auto-doubling of enemies by code, inside maps.

**Comment by Tormentor667** [comment]
Keys could be shared in all player inventories as well in my opinion.

Enemy AI attack the first player in sight or the first player they get damage from (or the player that is the highest treat) - common Doom AI

**Comment by DoomJedi** [comment]
Any progress?

**Comment by Tormentor667** [comment]
Well I guess it’s not high priority currently for @AFADoomer when it comes to the things we need to prepare for 3.1

**Comment by AFADoomer** [comment]
Wait, what am I supposed to be doing?  Shared inventory and/or keys?

**Comment by Tormentor667** [comment]
Shared keys and puzzle items actually would make sense but it is still open for discussion . What do you think?

**Comment by Username-N00b-is-not-available** [comment]
![coop](https://user-images.githubusercontent.com/59409713/121903547-18513980-cd31-11eb-8256-9ca22712e1d3.png)

I have just launched two copies of BoA on one machine into a network game as suggested [here](https://zdoom.org/wiki/Multiplayer#Technical_details) (used ```boa -host 2 +map c2m3 +skill 0``` and ```boa -join 192.168.X.X``` with my current local IP address), and was pleasantly surprised: there are little problems with the core 'gameplay loop'. As @Ozymandias81 stated in https://github.com/Realm667/WolfenDoom/issues/623#issuecomment-840553438 and https://github.com/Realm667/WolfenDoom/issues/623#issuecomment-840728174, there sure will be problems on specific maps currently, but at least now it seems to works _adequately_.
Here are some minor problems I spotted for now after playing a few minutes (let's leave fixing them to time **after** 3.1 has been released):

- [ ] Players can't use kicks.
- [ ] Gold items do not disappear on pickup, so the player can pick them up indefinitely and 35 times per second, note the upper left corner of the image.
- [ ] Secret hint sheets do not work.
- [ ] Red directional damage indicator at the centre of the screen shows up for other players, too.

***
I'd say that multiplayer support could definitely make a good addition for v3.2 (3.3? 3.8..?) in the future (but will mean a lot of work obviously, with testing as the main problem).

**Comment by Tormentor667** [comment]
I think this sounds promising and defintitlely solvable

**Comment by Username-N00b-is-not-available** [comment]
There is also such a thing as deathmatch, and it could be even easier to complete for us — if we manage to find map parts which could make good DM maps. For example, here is what can be done with the C2M4 town (WAD in zip): [boa_dm.zip](https://github.com/Realm667/WolfenDoom/files/6657881/boa_dm.zip)
I tested it with zcajun bots, and it actually seems quite enjoyable, here are some problems/notes I figured.
- [ ] (probably the most important one) After the first death and respawn, the playerpawn's state becomes somehow broken; Blazkowicz starts running instead of walking, and the red haze does not fade and keeps obstructing the player view. This does not happen in coop with bots, only in deathmatch.
- If we are to release DM maps with some kind of bots, I suggest using something like https://forum.zdoom.org/viewtopic.php?f=43&t=64527 (have not worked with it yet, just googled it), because plain zcajun ones are too dumb — their implementation has not changed for around 20 years AFAIK. TDBots seem to be 'customizable', but we would need to modify weapon code slightly for them to work, specifically to insert some inventory checks (I would rather search for a workaround/try to rewrite the corresponding ZScript part of the bots). Stock bots are bad at navigating levels where you need to jump a lot (and are completely unaware of @AFADoomer's ladders), get stuck at places near weapons while trying to get them, and have problems with shooting single-shot weapons such as Luger or Klein Nebelwerfer.
- Some default skins directly based on actor sprites we currently have could be included (Blazkowicz as default, Douglas, Ryan, Ascher, allied paratroopers, Soviets). This will mostly be renaming work (unless, of course, these actors have less frames than BoAPlayer...)
- We'd need to find small enough parts of our maps that could be cut and converted into deathmatch areas. These should be mostly open or non-linear so the players do not end up running through one corridor back and forth. Aside from C2M4, I would try converting some areas of C1M2 (the very first complex), C2M5_B (large areas from the centre of the map, game with big weapons), C3M0_A (mansion, cinema or opera), C3M0_B (most of the complex)... Please suggest other maps. Obviously I do not mean that we should just cut the part out in editor, put player starts and call the chunk a deathmatch map, but rather modify it, open most of the doors, add some passages, distribute item pickups, etc.
- Performance: GZDoom only sends player ticcmds, so BoA complexity will not encumber the connection. In any case, I'd remove most of decorative actors from the map chunks.

**Comment by DoomJedi** [comment]
You can use also some scrapped art, like snow russian etc.

**Comment by Username-N00b-is-not-available** [comment]
Ah, we don't have crouching sprites for them...

**Comment by DoomJedi** [comment]
No russian has crouching sprites as they were allies, those were made only for enemies.

Crouching sprites are very easy to do, in general - and especially as they are just recolor of existing frames for other soldiers.
Fuller rolling animation - is a bit different thing.

**Comment by Username-N00b-is-not-available** [comment]
OK, if so I hope you don't mind if I recolor them (including Douglas, Ascher and allied marines) myself when I have time.
By the way, it seems that default ZDoom skin format does not work with BoA's player class, I should not forget to do some research in engine sources.

**Comment by Tormentor667** [comment]
I really like where this is leading to, but I'd really postpone that for now until everything for the v3.1 release is finished and prepared. I think we (especially @AFADoomer ) have already enough to currently take care of atm. But I am eager to continue here.

**Comment by Username-N00b-is-not-available** [comment]
Yes, this must definitely wait until we release 3.1, nothing urgent obviously. I was just collecting here the issues I noticed, so that they won't get forgotten when we have time to deal with them.

**Comment by Tormentor667** [comment]
**The king is dead, long live the king.**

Time to talk about the cooperative mode for the v3.2 release I guess.

**Comment by Username-N00b-is-not-available** [comment]
Some further comments.
For some reason co-op (two clients on one machine) desynchronizes for me near the beginning on C2M5_B. I didn't manage to track down the source of the problem, but was able to reproduce it a few times...
In deathmatch with bots, the red blood pool which appears under the player on his death sometimes carries over to the respawned player pawn.
Is there a way to make the player skins feature (S_SKIN) work with our player class, by the way? Does anybody know where the code is located in GZDoom?

**Comment by Tormentor667** [comment]
I noticed the red blood pool issue also when resurrecting myself after dying a number of times. It happens occationally.

Regarding S_SKIN, no idea, it's a Skulltag/Zandronum feature, isn't it?

**Comment by ColerXXX** [comment]
> B.J. Blazkowicz tried to puke script "QuickKick" (
> )

I think it's because it detects it as a cheat, same as when trying to show objectives for the little I tested, try to enable cheats with "sv_cheats true" command

**Comment by Tormentor667** [comment]
> I think it's because it detects it as a cheat, same as when trying to show objectives for the little I tested, try to enable cheats with "sv_cheats true" command

I wonder what can be done about this issue in general? Any idea or experiences?

**Regarding Coop**, I would like to prepare a coop beta version soon that I will make accessable for players out there. The question right now is simply: How do we prepare the coop version in general? With a launcher? Or with the explanation on how to start it command-line? And how do we deal with the boss enemies?

I'd be happy to start a discussion here with all of you @Ozymandias81 @AFADoomer @DoomJedi @doomjuan @MuadDib90 @Username-N00b-is-not-available

**Comment by MuadDib90** [comment]
I would welcome the launcher for sure, but I am also capable of hosting/joining games using command lines. (Well, but if the launcher would be too complicated to do only for beta testing, let's stick with command line.)

And what exactly do you mean considering bosses? (I am saying it from the laic position.) It does not matter who - in coop - will slay him, or does it? When his death does not trigger anything, there should be no problem. If so, it gets triggered anyway...

**Comment by ColerXXX** [comment]
For me also is ok with batch command lines! I think it would be enough just make it asking the ip to connect at when running the .bat file, 
thanks for your interest in coop mode! :)
